### PR TITLE
[VEN-1019][VEN-1022]Added loops limit.

### PR DESCRIPTION
**Pull request by defcon022**
## Description

Resolves #

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Comptroller.sol [hunk 1]
```diff
@@ -120,13 +123,21 @@ contract Comptroller is Ownable2StepUpgradeable, ComptrollerStorage, Comptroller
     /// @notice Thrown if the borrow cap is exceeded
     error BorrowCapExceeded(address market, uint256 cap);
 
+    /// @notice Thrown an error on maxLoopsLimit execcds for any for loop
```
Comment (chechu) [inline contracts/Comptroller.sol:126]:
```suggestion
    /// @notice Thrown an error on maxLoopsLimit exceeds for any loop
```

File: contracts/Rewards/RewardsDistributor.sol [hunk 2]
```diff
@@ -93,6 +96,12 @@ contract RewardsDistributor is ExponentialNoError, Ownable2StepUpgradeable {
     /// @notice Emitted when a reward for contributor is updated
     event ContributorRewardsUpdated(address contributor, uint256 rewardAccrued);
 
+    /// @notice Emitted when max loops limit is set
+    event MaxLoopsLimitUpdated(uint256 oldMaxLoopsLimit, uint256 newmaxLoopsLimit);
+
+    /// @notice Thrown an error on maxLoopsLimit execcds for any for loop
```
Comment (chechu) [inline contracts/Rewards/RewardsDistributor.sol:102]:
```suggestion
    /// @notice Thrown an error on maxLoopsLimit exceeds for any loop
```

File: contracts/RiskFund/RiskFund.sol [hunk 3]
```diff
@@ -48,6 +51,12 @@ contract RiskFund is Ownable2StepUpgradeable, ExponentialNoError, ReserveHelpers
     /// @notice Emitted when reserves are transferred for auction
     event TransferredReserveForAuction(address comptroller, uint256 amount);
 
+    /// @notice Emitted when max loops limit is set
+    event MaxLoopsLimitUpdated(uint256 oldMaxLoopsLimit, uint256 newmaxLoopsLimit);
+
+    /// @notice Thrown an error on maxLoopsLimit execcds for any for loop
```
Comment (chechu) [inline contracts/RiskFund/RiskFund.sol:57]:
```suggestion
    /// @notice Thrown an error on maxLoopsLimit exceeds for any loop
```

**Review by chechu — COMMENTED** [review]
Would it make sense to define a parent abstract class with these items?
* event MaxLoopsLimitUpdated
* error MaxLoopsLimitExceeded
* store variable `uint256 public maxLoopsLimit;`
* functions `setMaxLoopsLimit` and `_setMaxLoopsLimit`
* function `ensureMaxLoops(uint len)`, that would check if `len > maxLoopsLimit`

I think this way we would simplify the integration of this feature in our code. What do you think?

**Review by chechu — COMMENTED** [review]
I miss some tests on the specific added feature:
* Setting max loops limit and check the emitted event
* Setting max loops limit, iterate on a bigger loop and check the expected error

File: contracts/Comptroller.sol (hunk 1 above)
Comment (defcon022) [reply contracts/Comptroller.sol:126]: Fixed.

File: contracts/Rewards/RewardsDistributor.sol (hunk 2 above)
Comment (defcon022) [reply contracts/Rewards/RewardsDistributor.sol:102]: Fixed.

File: contracts/RiskFund/RiskFund.sol (hunk 3 above)
Comment (defcon022) [reply contracts/RiskFund/RiskFund.sol:57]: Fixed.

**Comment by defcon022** [comment]
> Would it make sense to define a parent abstract class with these items?
> 
> * event MaxLoopsLimitUpdated
> * error MaxLoopsLimitExceeded
> * store variable `uint256 public maxLoopsLimit;`
> * functions `setMaxLoopsLimit` and `_setMaxLoopsLimit`
> * function `ensureMaxLoops(uint len)`, that would check if `len > maxLoopsLimit`
> 
> I think this way we would simplify the integration of this feature in our code. What do you think?

I have created a seperate abstract contract for that.

**Comment by defcon022** [comment]
> I miss some tests on the specific added feature:
> 
> * Setting max loops limit and check the emitted event
> * Setting max loops limit, iterate on a bigger loop and check the expected error

Added.

File: tests/hardhat/MaxLoopsLimitHelper.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    maxLoopsLimitHelperFactory = await smock.mock<HarnessMaxLoopsLimitHelper__factory>("HarnessMaxLoopsLimitHelper");
+    maxLoopsLimitHelper = await maxLoopsLimitHelperFactory.deploy();
+  });
+
+  it("Set the maxLoopsLimit cap", async () => {
+    const result = await maxLoopsLimitHelper.setMaxLoopsLimit(150);
+    await expect(result).to.emit(maxLoopsLimitHelper, "MaxLoopsLimitUpdated").withArgs(0, 150);
+  });
+
+  it.only("Revert on exceeding the max loop limit", async () => {
```
Comment (chechu) [inline tests/hardhat/MaxLoopsLimitHelper.ts:23]:
```suggestion
  it("Revert on exceeding the max loop limit", async () => {
```

File: contracts/MaxLoopsLimitHelper.sol [hunk 5]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.13;
+
+abstract contract MaxLoopsLimitHelper {
+    // Limit for the loops to avoid the DOS
+    uint256 public maxLoopsLimit;
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/MaxLoopsLimitHelper.sol:6]: Added.

File: tests/hardhat/MaxLoopsLimitHelper.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { MockContract, MockContractFactory, smock } from "@defi-wonderland/smock";
+import chai from "chai";
+
+import { HarnessMaxLoopsLimitHelper, HarnessMaxLoopsLimitHelper__factory } from "../../typechain";
+
+const { expect } = chai;
+chai.use(smock.matchers);
+
+describe.only("MaxLoopsLimit: tests", () => {
```
[the remark this answers is not in the document]
Comment (defcon022) [reply tests/hardhat/MaxLoopsLimitHelper.ts:9]: Oops, fixed.

File: tests/hardhat/MaxLoopsLimitHelper.ts (hunk 4 above)
Comment (defcon022) [reply tests/hardhat/MaxLoopsLimitHelper.ts:23]: Oops, fixed.

**Review by chechu — APPROVED** [review]
Thank you! Good PR

**Review by narayanprusty — APPROVED** [review]

**Pull request merged**